toolkit/tools/imagegen/configuration/packagerepo.go

```diff
@@ -0,0 +1,180 @@
+
+// PackageRepo defines the RPM repo to pull packages from during the installation
+// or after the system is installed. The "Install" option indicates that the provided
+// repository configuration will be saved in the installed system if specified, and only
+// available during the installation process if not
+type PackageRepo struct {
+	Name    string `json:"Name"`
+	BaseUrl string `json:"BaseUrl"`
+	Install bool   `json:"Install"`
+}
+
+// UnmarshalJSON Unmarshals a PackageRepo entry
+func (p *PackageRepo) UnmarshalJSON(b []byte) (err error) {
+	// Use an intermediate type which will use the default JSON unmarshal implementation
+	type IntermediateTypePackageRepo PackageRepo
+	err = json.Unmarshal(b, (*IntermediateTypePackageRepo)(p))
+	if err != nil {
+		return fmt.Errorf("failed to parse [PackageRepo]: %w", err)
+	}
+
+	// Now validate the resulting unmarshaled object
+	err = p.IsValid()
+	if err != nil {
+		return fmt.Errorf("failed to parse [PackageRepo]: %w", err)
+	}
+	return
+}
+
+// IsValid returns an error if the PackageRepo struct is not valid
+func (p *PackageRepo) IsValid() (err error) {
+	err = p.nameIsValid()
+	if err != nil {
+		return
+	}
+
+	err = p.repoUrlIsValid()
+	if err != nil {
+		return
+	}
+
+	return
+}
+
+// UpdatePackageRepo creates additional repo files specified by image configuration
+// and returns error if the operation fails
+func UpdatePackageRepo(installChroot *safechroot.Chroot, config SystemConfig) (err error) {
+	const (
+		repoFileDir   = "/etc/yum.repos.d/"
+		localRepoFile = "/etc/yum.repos.d/mariner-iso.repo"
+		squashErrors  = false
+	)
+
+	if exists, ferr := file.DirExists(repoFileDir); ferr != nil {
+		logger.Log.Errorf("Error accessing repo file directory.")
+		err = ferr
+		return
+	} else if !exists {
+		err = fmt.Errorf("%s: no such directory", repoFileDir)
+		return
+	}
+
+	// Remove mariner-iso.repo
+	os.Remove(localRepoFile)
+
+	// Loop through the PackageRepos field to determine if any customized package repos are specified.
+	// If specified, create new repo files for them
+	for _, packageRepo := range config.PackageRepos {
+		err = createCustomPackageRepo(installChroot, packageRepo, repoFileDir)
+		if err != nil {
+			return
+		}
+	}
+
+	// It is possible that network access may not be up at this point,
+	// so check network access
+	err = network.CheckNetworkAccess()
+	return
+}
+
+// nameIsValid returns an error if the package repo name is empty
+func (p *PackageRepo) nameIsValid() (err error) {
+	if strings.TrimSpace(p.Name) == "" {
+		return fmt.Errorf("invalid value for package repo name (%s), name cannot be empty", p.Name)
+	}
+	return
+}
+
+// repoUrlIsValid returns an error if the package url is invalid
+func (p *PackageRepo) repoUrlIsValid() (err error) {
+	if strings.TrimSpace(p.BaseUrl) == "" {
+		return fmt.Errorf("invalid value for package repo URL (%s), URL cannot be empty", p.BaseUrl)
+	}
+
+	_, err = url.ParseRequestURI(p.BaseUrl)
+	if err != nil {
+		return
+	}
+	return
```

Comment:
Nit: since this is the last call, we can also leave with a simple `return` regardless of the error. Having said that, we would benefit from adding another error log here to make it easier to find the place where the code failed.

Reply:
Could you still add that additional log, please?